prometheus_juju_exporter/collector.py

```diff
@@ -1,41 +1,265 @@
-        """Create a dict of gauge instances."""
-        if gauge_name not in self.metrics:
-            self.logger.debug("Creating Gauge %s", gauge_name)
-            self.metrics[gauge_name] = Gauge(
-                gauge_name, gauge_desc, labelnames=labels, registry=self._registry
+        self.logger = getLogger(__name__)
+        self.controller = Controller()
+        self.model = Model()
+        self.data = {}
+        self.currently_cached_labels = {}
+        self.previously_cached_labels = {}
+        self.logger.debug("Collector initialized")
+        self.prefixes = [
+            "52:54:00",
+            "fa:16:3e",
+            "06:f1:3a",
+            "00:0d:3a",
+            "00:50:56",
+            "fa:16:3e",
+        ]
+
+    def refresh_cache(self, gauge_name, gauge_desc, labels):
+        """Refresh instances for each collection job.
+
+        :param str gauge_name: the name of the gauge
+        :param str gauge_desc: the description of the gauge
+        :param List[str] labels: the label set of the gauge
+        """
+        self.data = {
+            gauge_name: {
+                "gauge_desc": gauge_desc,
+                "labels": labels,
+                "labelvalues_update": [],
+                "labelvalues_remove": [],
+            }
+        }
+
+        self.previously_cached_labels = self.currently_cached_labels.copy()
+        self.currently_cached_labels = dict()
+
+        self.controller = Controller()
+        self.model = Model()
+
+    async def _connect_controller(self, endpoint, username, password, cacert):
+        """Connect to a controller via its endpoint.
+
+        :param str endpoint: the hostname:port endpoint of the controller
+            to connect to.
+        :param str username: the username for controller-local users
+        :param str password: the password for controller-local users
+        :param str cacert: the CA certificate of the controller
+            (PEM formatted)
+        """
+        if not self.controller.is_connected():
+            await self.controller.connect(
+                endpoint=endpoint, username=username, password=password, cacert=cacert
             )
 
+    async def _get_models(self, endpoint, username, password, cacert):
+        """Get a list of all models under a controller.
+
+        :param str endpoint: the hostname:port endpoint of the controller
+            to connect to.
+        :param str username: the username for controller-local users
+        :param str password: the password for controller-local users
+        :param str cacert: the CA certificate of the controller
+            (PEM formatted)
+        :return: str model_uuids: the uuids of all models under the controller
+        """
+        await self._connect_controller(
+            endpoint=endpoint, username=username, password=password, cacert=cacert
+        )
+        model_uuids = await self.controller.model_uuids()
+        await self.controller.disconnect()
```

Comment:
I think that we should use logic like this
```python
async def _connect_controller(...):
    ...
    self.controller = Controller(max_frame_size= 6**24)
    self.connect(endpoint=endpoint, username=username, password=password, cacert=cacert)
    ...

async def _get_models(self):
    model_uuids = await self.controller.model_uuids()
    return model_uuids

async def _get_machines_in_model(self, uuid):
    model = await self.controller.get_model(uuid)
    status = await model.get_status()
    return status["machines"]

async def get_stats(self):
    ...
    await self._connect_controller(endpoint=endpoint, username=username, password=password, cacert=cacert)
    model_uuids = await self._get_models()
    ...
    for name, uuid_ in model_uuids.items():
        self.logger.debug("Checking model '%s'...", name)
        machines = await self._get_machines_in_model(uuid_)
        ...
    await self.controller.disconnect()
```
Like this you connecting only once to the controller, because `model.connect` is actually connecting to the same controller again and again.

Reply:
Maybe also we can add
```python
try:
    await self._connect_controller(endpoint=endpoint, username=username, password=password, cacert=cacert)
    ...
finally:
    await self.controller.disconnect()
```
to ensure that we disconnect from controller. Or define `self._connect_controller` function as context_manager.